doc: Upgrade guidance for existing `mongodbatlas_advanced_cluster` with pre and post ISS (#3204)
mongodb/terraform-provider-mongodbatlas: docs/resources/advanced_cluster (preview provider 2.0.0).md

## How to migrate

If you're currently utilizing `mongodbatlas_cluster`, see our [Migration Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/cluster-to-advanced-cluster-migration-guide).

> If you're currently utilizing `mongodbatlas_advanced_cluster`, you should follow these steps in order:

Reviewer: ```suggestion
If you're currently utilizing `mongodbatlas_advanced_cluster`, you can proactively address the upcoming breaking changes that will affect all `mongodbatlas_advanced_cluster` resources when the next major provider version, 2.0.0, is released (timeline yet to be announced). However, it's not necessary to do this right now. If you decide to go ahead, we recommend to follow these steps in order:
```

> If you're currently utilizing `mongodbatlas_advanced_cluster`, you can proactively address the upcoming breaking changes that will affect all `mongodbatlas_advanced_cluster` resources when the next major provider version, 2.0.0, is released (timeline yet to be announced). However, it's not necessary to do this right now. If you decide to go ahead, we recommend to follow these steps in order:

Reviewer 2: ```suggestion
If you're currently using `mongodbatlas_advanced_cluster`, you can proactively address the upcoming breaking changes that will affect all `mongodbatlas_advanced_cluster` resources when the next major provider version, 2.0.0, is released (timeline yet to be announced). To avoid these breaking changes, we recommend that you follow these steps in order:
```

> 1. If you are using the deprecated sharding configuration (with `num_shards`), you must first migrate to the new Independent Shard Scaling schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. This step is required before enabling the Preview for MongoDB Atlas Provider 2.0.0.

Reviewer: can you add a link to the num_shards field and a link to the Independent Shard Scaling schema? Maybe @AgustinBettati could help with what best link we have there

Reviewer 3: I believe the guide that is referenced (https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) is the best point for getting details of the ISS schema.

Author: The ISS link is already in the following sentence, so I wouldn't add it again in `Independent Shard Scaling`. On the num_shards, there is the https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/advanced_cluster#num_shards-1 but I think the already linked guide does a much better job at explaining num_shards effect and how to migrate to the new schema

Reviewer: Do you think we should pick only one? 
On my side I do see the value in linking num_shards mention here to its documentation section because it gives a good explanation of what it is, that it is deprecated and the upgrade guide. We just got a HELP ticket last week about clarity on asymmetric and symmetric AS and the relation with this field.

Author: no strong opinion, added your suggestions in [bdac7b6](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3204/commits/bdac7b6f1cec64502542253a2a9a96ecfd1acf48)

> 1. If you are using the deprecated sharding configuration (with `num_shards`), you must first migrate to the new Independent Shard Scaling schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. This step is required before enabling the Preview for MongoDB Atlas Provider 2.0.0.

Reviewer: ```suggestion
1. If you are using the deprecated sharding configuration (with `num_shards`), you should first migrate to the new Independent Shard Scaling schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. This step is recommended before enabling the Preview for MongoDB Atlas Provider 2.0.0.
```

> 2. Once your configuration uses the new sharding schema (or if you were already using it), you can proceed to enable the Preview for MongoDB Atlas Provider 2.0.0 by following these steps:

Reviewer: when we talk about the PReview, let's link it to https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/advanced_cluster%2520%2528preview%2520provider%25202.0.0%2529

Author: Does it make sense to link into the same file?

Reviewer: Fair point, my mind was on another doc.

> 1. If you are using the deprecated sharding configuration (with `num_shards`), you should first migrate to the new Independent Shard Scaling schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. This step is required before enabling the Preview for MongoDB Atlas Provider 2.0.0.

Reviewer: ```suggestion
1. If you are using the deprecated sharding configuration (with `num_shards`), you should first migrate to the new Independent Shard Scaling schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. This step is recommended before enabling the Preview for MongoDB Atlas Provider 2.0.0.
```

Reviewer: I made this suggestion before but not sure why it didn't go through

> 2. Once your configuration uses the new sharding schema (or if you were already using it), you can proceed to enable the Preview for MongoDB Atlas Provider 2.0.0 by following these steps:

Reviewer 3: ```suggestion
2. You can proceed to enable the Preview for MongoDB Atlas Provider 2.0.0 by following these steps:
```
following similar suggestion from Marco that it is not a required but recommended step

> 1. If you are using the deprecated sharding configuration (with [`num_shards`](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/advanced_cluster#num_shards-1)), you should first migrate to the new [Independent Shard Scaling]((https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema)) schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. This step is recommended before enabling the Preview for MongoDB Atlas Provider 2.0.0.

Reviewer 2: I think the last sentence is a bit redundant. It's clear from the lead in that these are the recommended steps. Removing this also makes this a much stronger recommendation
```suggestion
1. If you are using the deprecated sharding configuration (with [`num_shards`](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/advanced_cluster#num_shards-1)), you should first migrate to the new [Independent Shard Scaling]((https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema)) schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. 
```

> 2. You can proceed to enable the Preview for MongoDB Atlas Provider 2.0.0 by following these steps:

Reviewer 2: ```suggestion
2. Enable the Preview for MongoDB Atlas Provider 2.0.0 by following these steps:
```

>   - Before starting, run `terraform plan` to make sure that there are no planned changes.

Reviewer 2: ```suggestion
  - Run `terraform plan` to make sure that there are no planned changes.
```

>   - Set environment variable `MONGODB_ATLAS_PREVIEW_PROVIDER_V2_ADVANCED_CLUSTER=true` in order to use the Preview for MongoDB Atlas Provider 2.0.0.

Reviewer 2: ```suggestion
  - Set the environment variable `MONGODB_ATLAS_PREVIEW_PROVIDER_V2_ADVANCED_CLUSTER=true` in order to use the Preview for MongoDB Atlas Provider 2.0.0.
```

  - Run `terraform plan` and you'll see errors as definition file hasn't been updated yet.
  - Apply definition changes explained on this page until there are no errors and no planned changes. **Important**: Don't apply until the plan is empty. If it shows other changes, you must update the `mongodbatlas_advanced_cluster` configuration until it matches the original configuration.
  - Run `terraform apply` to apply the changes. Although there are no plan changes shown to the user, the `mongodbatlas_advanced_cluster` state will be updated to support the Preview for MongoDB Atlas Provider 2.0.0. **Important**: The state file will be updated to the new format and the old format will not be supported anymore. Before doing any migration create a backup of your [Terraform state file](https://developer.hashicorp.com/terraform/cli/commands/state).

> -> **IMPORTANT:** Do not enable the Preview for MongoDB Atlas Provider 2.0.0 if you are still using the deprecated sharding configuration. This would require you to perform two separate migrations when version 2.0.0 is released: one to update to the new configuration format and another to migrate to the new sharding schema. Instead, migrate to the new sharding schema first, then enable the preview.

Reviewer: ```suggestion
-> **IMPORTANT:** If you migrate to the [Preview for MongoDB Atlas Provider 2.0.0](ADD LINK HERE) while still using the [deprecated sharding configuration](ADD LINK HERE), you will be required to perform the migration to the new Independent Shard Scaling schema when version 2.0.0 is released.
```

Reviewer 3: Would also link https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema when mentioning Independent Shard Scaling schema.

Author: addressed in [8887633](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3204/commits/8887633c33491c493cfcbe38856466371676a5dc)

## How to migrate

If you're currently utilizing `mongodbatlas_cluster`, see our [Migration Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/cluster-to-advanced-cluster-migration-guide).

> If you're currently utilizing `mongodbatlas_advanced_cluster`, while it's not necessary to do this right now, you can proactively address the upcoming breaking changes that will affect all `mongodbatlas_advanced_cluster` resources when the next major provider version, 2.0.0, is released (timeline yet to be announced).
> If you decide to go ahead, we recommend to follow these steps in order:

Reviewer 4: ```suggestion
If you're currently utilizing `mongodbatlas_advanced_cluster`, you may also proactively address the upcoming breaking changes that will affect all `mongodbatlas_advanced_cluster` resources when the next major provider version, 2.0.0, is released (timeline yet to be announced).
If you decide to go ahead, we recommend to follow these steps in order:
```

Reviewer 4: nit in wording. Not necessary begs the question of why when you just want to be clear on the conditionality

1. If you are using the deprecated sharding configuration (with [`num_shards`](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/advanced_cluster#num_shards-1)), you should first migrate to the new [Independent Shard Scaling]((https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema)) schema. See our [Advanced Cluster New Sharding Configurations Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema) for details. 

> ~> **IMPORTANT:** Before doing any migration, create a backup of your [Terraform state file](https://developer.hashicorp.com/terraform/cli/commands/state). The state file will be updated to the new format and the old format will not be supported anymore.

Reviewer: @oarbusi I've replaced -> with ~> because I think that's the correct syntax, but I need to sanity check this with you

Author: I think your change is correct. -> is the blue box that we usually use for "Notes". ~> is the yellow box more like a warning(see https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/encryption_at_rest)

> ~> **IMPORTANT:** Before doing any migration, create a backup of your [Terraform state file](https://developer.hashicorp.com/terraform/cli/commands/state). The state file will be updated to the new format and the old format will not be supported anymore.

Reviewer 4: ```suggestion
~> **IMPORTANT:** Before doing any migration, create a backup of your [Terraform state file](https://developer.hashicorp.com/terraform/cli/commands/state). The state file will be updated to the new format and the old format will no longer be supported.
```

2. Enable the Preview for MongoDB Atlas Provider 2.0.0 by following these steps:
  - Run `terraform plan` to make sure that there are no planned changes.
  - Set the environment variable `MONGODB_ATLAS_PREVIEW_PROVIDER_V2_ADVANCED_CLUSTER=true` in order to use the Preview for MongoDB Atlas Provider 2.0.0.
  - If you run `terraform plan` again, you'll see syntax errors: this is expected since the definition file hasn't been updated yet using the latest schema.

>   - At this point, you can apply definition changes explained on this page until there are no errors and no planned changes. **Important**: Don't apply until the plan is empty. If it shows other changes, you must update the `mongodbatlas_advanced_cluster` configuration until it matches the original configuration.

Reviewer 4: > definition changes explained on this page
Nit: 'this' is vague/unclear. Link to the earlier configuration changes to be clear which changes are required.

Reviewer: good point, done
